### PR TITLE
[not ready] Parallelize more: nested parallelism, rely on task scheduler

### PR DESCRIPTION
**Pull request by daniel-j-h**
This Pull Request tracks the progress. Do not merge yet.

The high level goal is to parallelize more using Intel TBB's scheduler for:

- nested parallelism
- not oversubscribing

Please read every commit's description, as they have detailed explanations attached. This pull request just tracks the overall progress.

[Here's](https://github.com/Project-OSRM/osrm-backend/commit/6b2bf4954aa87402e3d64c0278b26c4a61e8e667) additional explanation on this, as in the following commit messages.

### CODE REVIEW

**Comment by TheMarex** [comment]
@daniel-j-h TBB should only be used in preprocessing - not in the server component. We rely on request-level parallelism in the server.

File: contractor/contractor.hpp [hunk 1]
```diff
@@ -411,7 +416,7 @@ class Contractor
                 contractor_graph.reset();
 
                 // create new graph
-                std::sort(new_edge_set.begin(), new_edge_set.end());
```
Comment (TheMarex) [inline contractor/contractor.hpp]: Good catch, this should definitely help.

File: contractor/contractor.hpp [hunk 2]
```diff
@@ -463,7 +468,7 @@ class Contractor
                 [&](const ThreadDataContainer::EnumerableThreadData::range_type &range)
                 {
                     for (auto &data : range)
-                        std::sort(data->inserted_edges.begin(), data->inserted_edges.end());
```
Comment (TheMarex) [inline contractor/contractor.hpp]: This could help - but definitely benchmark this.

File: contractor/contractor.hpp [hunk 3]
```diff
@@ -861,7 +867,7 @@ class Contractor
             }
         }
         // eliminate duplicate entries ( forward + backward edges )
-        std::sort(neighbours.begin(), neighbours.end());
```
Comment (TheMarex) [inline contractor/contractor.hpp]: This is a hot code path - will probably be slower if you invoke TBB.

File: contractor/contractor.hpp [hunk 4]
```diff
@@ -890,7 +896,7 @@ class Contractor
             node_data[u].depth = (std::max)(node_data[node].depth + 1, node_data[u].depth);
         }
         // eliminate duplicate entries ( forward + backward edges )
-        std::sort(neighbours.begin(), neighbours.end());
```
Comment (TheMarex) [inline contractor/contractor.hpp]: This is also a hot code path.

File: contractor/contractor.hpp [hunk 5]
```diff
@@ -933,7 +939,7 @@ class Contractor
             neighbours.push_back(target);
         }
 
-        std::sort(neighbours.begin(), neighbours.end());
```
Comment (TheMarex) [inline contractor/contractor.hpp]: Also a hot code path.

File: routing_algorithms/alternative_path.hpp [hunk 6]
```diff
@@ -287,7 +292,7 @@ class AlternativeRouting final
                 ranked_candidates_list.emplace_back(node, length_of_via_path, sharing_of_via_path);
             }
         }
-        std::sort(ranked_candidates_list.begin(), ranked_candidates_list.end());
```
Comment (TheMarex) [inline routing_algorithms/alternative_path.hpp]: This is server side.

File: routing_algorithms/direct_shortest_path.hpp [hunk 7]
```diff
@@ -161,8 +167,8 @@ class DirectShortestPathRouting final
             {
                 return lhs.first < rhs.first || (lhs.first == rhs.first && lhs.second < rhs.second);
             };
-        std::sort(forward_entry_points.begin(), forward_entry_points.end(), entry_point_comparator);
```
Comment (TheMarex) [inline routing_algorithms/direct_shortest_path.hpp]: This is server side.

File: util/container.hpp [hunk 8]
```diff
@@ -49,7 +52,7 @@ constexpr bool has_resize_method(...) noexcept { return false; }
 
 template <typename Container> void sort_unique_resize(Container &vector) noexcept
 {
-    std::sort(std::begin(vector), std::end(vector));
```
Comment (TheMarex) [inline util/container.hpp]: This _might_ be used by the server.

**Comment by TheMarex** [comment]
If you remove calls inside hot code paths, we can merge this.

**Comment by daniel-j-h** [comment]
Tested on Germany and Sweden, calls to the TBB function on the hot path did not had any measurable impact. Removed them nevertheless.

From a theoretical standpoint, TBB probably has a guard that checks for the collection's siye before dispatching and doing small sizes in serial, which explains this behavior.

Waiting on Travis and then merging this.

**Pull request merged**